snuba/web/split.py

```diff
@@ -1,34 +1,153 @@
 import copy
 import math
-from datetime import timedelta
-from typing import Any, List, Optional, Union
+from dataclasses import replace
+from datetime import datetime, timedelta
+from typing import Any, List, Optional, Sequence, Tuple, Union
 
-from snuba import state, util
+from snuba import environment, state, util
 from snuba.clickhouse.query import Query
 from snuba.datasets.plans.split_strategy import QuerySplitStrategy, SplitQueryRunner
+from snuba.query.conditions import (
+    OPERATOR_TO_FUNCTION,
+    BooleanFunctions,
+    binary_condition,
+    in_condition,
+    is_binary_condition,
+)
+from snuba.query.dsl import literals_tuple
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
 from snuba.request.request_settings import RequestSettings
 from snuba.util import is_condition
+from snuba.utils.metrics.backends.wrapper import MetricsWrapper
 from snuba.web import QueryResult
 
+metrics = MetricsWrapper(environment.metrics, "query.splitter")
+
 # Every time we find zero results for a given step, expand the search window by
 # this factor. Based on the assumption that the initial window is 2 hours, the
 # worst case (there are 0 results in the database) would have us making 4
 # queries before hitting the 90d limit (2+20+200+2000 hours == 92 days).
 STEP_GROWTH = 10
 
 
-def _identify_condition(condition: Any, field: str, operand: str) -> bool:
-    return is_condition(condition) and condition[0] == field and condition[1] == operand
+def _identify_condition(condition: Any, field: str, operator: str) -> bool:
+    return (
+        is_condition(condition) and condition[0] == field and condition[1] == operator
+    )
+
+
+def _get_time_range(
+    query: Query, timestamp_field: str
+) -> Tuple[Optional[Literal], Optional[Literal]]:
+    """
+    Finds the time range for this query. Which means, finds the timestamp condition
+    with the highest datetime literal and the timestamp condition with the smallest
+    and returns the interval in the form of a tuple of Literals. It only looks into
+    first level AND conditions.
+
+    TODO: Consider making this part of the AST manipulation api if there are more use
+    cases. It would require managing a few more corner cases for being part of the api.
+    """
+
+    def get_first_level_conditions(exp: Expression) -> Sequence[Expression]:
+        """
+        Traverse the tree as long as it runs into AND conditions and return all the
+        top level conditions.
+        """
+        if isinstance(exp, FunctionCall) and is_binary_condition(
+            exp, BooleanFunctions.AND
+        ):
+            return [
+                *get_first_level_conditions(exp.parameters[0]),
+                *get_first_level_conditions(exp.parameters[1]),
```

Comment:
Is this implying that only two conditions can be ANDed together at a time? Why not more?

Reply:
In the AST conditions are a tree of binary expressions like
```
          AND
         /     \           
     AND         C 
    /     \
A         B
```

So each FunctionCall representing an AND is binary. But creating a tree of binary conditions allow us to have multiple conditions in AND together.
That is why we execute this recursively until we find a node in the tree that does not represent an AND
Does this answer your question ?

Reply:
It does. I guess I didn't quite get the reason we need to limit these conditions to having 2 nodes in the AST though?

Reply:
A few reasons:
the most important to me is that when dealing with an AST we should reduce the amount of semantically equivalent operations as much as possible and represent the AST with as few concept as possible. This makes any query processing simpler since the processor code does not have to bother with multiple different ways to express the same structure.
forcing boolean functions to be binary (except for NOT) is the simplest possible form with the smallest amount of ambiguity.
If we allow AND and OR to have an arbitrary number of children, it would increase a lot the possible ways to express the same condition, examples:
```
AND
  a,
  b,
  AND
     c,
     d,
  e
```
is the same as 
```
AND 
   a
   b 
   c
   d
   e
```
as well as
```
AND
   AND
      AND
        d
        a
    c
b
```

Then it is up to the query processor to make sense of all those different structures.
Said that there are still different ways to express the same condition
```
AND
  a
  AND
     b
     c
```
is the same as
```
AND
  AND
     a
      b
c
```

But the possible number of permutations is a lot smaller and more manageable


The second reason is simplicity in translating and formatting the query:
we represent conditions as functions, so does clickhouse,  so we. do not need to do any extra processing to turn those function back into conditions, we just send them as function calls to clickhouse. And Clickhouse conditions are binary functions.

Reply:
Ok, thanks for the explanation.